Post for micro-frontends in Angular with shared auth state (#1161)
oktadeveloper/okta-blog: _source/_posts/2022-05-17-angular-microfrontend-auth.md

{% img blog/angular-microfrontend-auth/example-site.jpg alt:"Image of an example e-commerce site selling balloons with a header, links to account & cart, and images of balloons to purchase" width:"800" %}{: .center-image }

You might have a shopping cart, account information for registered users, past orders, payment options, etc. You might be able to further categorize these features into domains, each of which could be a separate micro frontend, also known as a `remote`. The collection of micro-frontend remotes is housed inside another website, the `host` of the web application.

So, your e-commerce site using micro frontends to decompose different functionality might look like this diagram, where the shopping cart and account features are in their separate routes within your Single Page Application (SPA):

{% img blog/angular-microfrontend-auth/example-site-diagram.jpg alt:"Image of the example balloon e-commerce site showing a breakdown of the views. The entire site is wrapped in a host named 'Shell', and micro-frontends for the 'Cart' and 'Account' links" width:"800" %}{: .center-image }

> You might be saying, "Micro-frontends sound cool, but managing the different frontends and orchestrating state across the micro frontends also sounds complicated." You're right. The concept of a micro frontend has been around for a few years and rolling your own micro-frontend implementation, shared state, and tools to support it _was_ quite an undertaking. However, micro frontends are now well supported with Webpack 5 and [Module Federation](https://webpack.js.org/concepts/module-federation/). Not all web apps require a micro-frontend architecture, but for those large, feature-rich web apps that have started to get unwieldy, the first-class support of micro frontends in our web tooling is definitely a plus.

Reviewer: ```suggestion
You might be saying, "Micro frontends sound cool, but managing the different frontends and orchestrating state across the micro frontends also sounds complicated." You're right. The concept of a micro frontend has been around for a few years and rolling your own micro-frontend implementation, shared state, and tools to support it _was_ quite an undertaking. However, micro frontends are now well supported with Webpack 5 and [Module Federation](https://webpack.js.org/concepts/module-federation/). Not all web apps require a micro-frontend architecture, but for those large, feature-rich web apps that have started to get unwieldy, the first-class support of micro frontends in our web tooling is definitely a plus.
```

      // },        
        
      // For hosts (please adjust)
      remotes: {
        "mfeBasket": "http://localhost:4201/remoteEntry.js",
      },

      shared: share({
        // ...important external libraries to share

        ...sharedMappings.getDescriptors()
      })
    }),
    sharedMappings.getPlugin()
  ],
};
```

In the `webpack.config.js` for `mfe-basket`, you'll see the path for `@shared` at the top of the file and the configuration to identify what to expose in the remote application.

The `shell` application serves on port 4200, and the `mfe-basket` application serves on port 4201. We can open up two terminals to run each application, or we can use the following npm script created for us by the schematic to add `@angular-architects/module-federation`:
 
```shell
npm run run:all
``` 

When you do so, you'll see both applications open in your browser and how they fit together in the `shell` application running on port 4200. Click the **Basket** button to navigate to a new route that displays the `BasketModule` in the `mfe-basket` application. The sign-in button doesn't work quite yet, but we'll get it going here next.

> **Note** - _Another option I could have used for the starter is a [Nx workspace](https://nx.dev/). Nx has great tooling and built-in support for building micro-frontends with Webpack and Module Federation. But I wanted to go minimalistic on the project tooling so you'd have a chance to dip your toes into some of the configuration requirements._

Reviewer: ```suggestion
**Note** - _Another option I could have used for the starter is a [Nx workspace](https://nx.dev/). Nx has great tooling and built-in support for building micro frontends with Webpack and Module Federation. But I wanted to go minimalistic on the project tooling so you'd have a chance to dip your toes into some of the configuration requirements._
```

## Micro-frontend state management

To share a library, you need to configure the library in the `webpack.config.js`. Let's start with `shell`. Open `projects/shell/src/webpack.config.js`.

There are two places to add shared code. One place is for code implementation within the project, and one is for shared external libraries. In this case, we can share the Okta external libraries as we didn't implement a service that wraps Okta's auth libraries, but I will point out both places.

First, we'll add the Okta libraries. Scroll down towards the bottom of the file to the `shared` property. You'll follow the same pattern as the `@angular` libraries already in the list and add the singleton instances of the two Okta libraries as shown in this snippet:

```js
shared: share({
  // other Angular libraries remain in the config. This is just a snippet
  "@angular/router": { singleton: true, strictVersion: true, requiredVersion: 'auto' },
  "@okta/okta-angular": { singleton: true, strictVersion: true, requiredVersion: 'auto' },
  "@okta/okta-auth-js": { singleton: true, strictVersion: true, requiredVersion: 'auto' },

  ...sharedMappings.getDescriptors()
})
```

> When you create a library within this project, like the basket service and project service in the starter code, you'd add the library to the `sharedMappings` array at the top of the `webpack.config.js` file. If you create a new library to wrap Okta's libraries, this is where you'd add it.

Reviewer: ```suggestion
When you create a library within this project, like the basket service and project service in the starter code, you add the library to the `sharedMappings` array at the top of the `webpack.config.js` file. If you create a new library to wrap Okta's libraries, this is where you'd add it.
```
verb tense alignment

> Now that you added the Okta libraries to the micro-frontend host, you need to also add them to the remotes that consume the dependencies. In our case, only the `mfe-profile` application uses Okta authenticated state information. Open `projects/mfe-profile/webpack.config.js`. Add the two Okta libraries to the `shared` property as you did for the `shell` application.

Reviewer: ```suggestion
Now that you've added the Okta libraries to the micro-frontend host, you need to also add them to the remotes that consume the dependencies. In our case, only the `mfe-profile` application uses Okta authenticated state information. Open `projects/mfe-profile/webpack.config.js`. Add the two Okta libraries to the `shared` property as you did for the `shell` application.
```
verb tense

Now, you should be able to run the project using `npm run run:all`, and the cupcake storefront should allow you to log in, see your profile, log out, and add items to your cupcake basket!

## Next steps

I hope you enjoyed this first post on creating an Angular micro-frontend site. We explored the capabilities of micro frontends and shared state between micro frontends using Webpack's Module Federation in Angular.  You can check out the completed code for this post in the [`local` branch in the @oktadev/okta-angular-microfrontend-example GitHub repo](https://github.com/oktadev/okta-angular-microfrontend-example/tree/local) by using the following command:

```shell
git clone --branch local https://github.com/oktadev/okta-angular-microfrontend-example.git
```

> Stay tuned for next time. I'll show how to prepare for deployment by transitioning to dynamic module loading and deploying the site to a free cloud provider.

Reviewer: ```suggestion
Stay tuned for part two. I'll show how to prepare for deployment by transitioning to dynamic module loading and deploying the site to a free cloud provider.
```
reminds the reader what "next time" you're referring to...

> ## Learn about Angular, OpenID Connect, micro-frontends and more

Reviewer: ```suggestion
## Learn about Angular, OpenID Connect, micro frontends, and more
```
serial comma, micro frontends as two words
